packages/devtools_app/lib/src/network/network_controller.dart

```diff
@@ -0,0 +1,204 @@
+    final httpEventIds = <String>{};
+    // Perform initial pass to find the IDs for the HTTP timeline events.
+    for (final TimelineEvent event in events) {
+      final json = event.toJson();
+      final id = json['id'];
+      if (id == null) {
+        continue;
+      }
+      // The start event for HTTP contains a filter key which we can use.
+      // Note: only HTTP client requests are currently logged to the timeline.
+      if ((!json['args'].containsKey('filterKey') ||
+              json['args']['filterKey'] != 'HTTP/client') &&
+          !outstandingRequestsMap.containsKey(id)) {
+        continue;
+      }
+      httpEventIds.add(id);
+    }
+
+    // Group all HTTP timeline events with the same ID.
+    final Map<String, List<Map>> httpEvents = {};
+    for (final event in events) {
+      final json = event.toJson();
+      final id = json['id'];
+      if (id == null) {
+        continue;
+      }
+      if (httpEventIds.contains(id)) {
+        if (!httpEvents.containsKey(id)) {
+          httpEvents[id] = [];
+        }
+        httpEvents[id].add(json);
+      }
+    }
+
+    // Build our list of network requests from the collected events.
+    for (final request in httpEvents.entries) {
+      final requestId = request.key;
+      final requestData = HttpRequestData.fromTimeline(
+          _timelineMicrosOffset, request.value);
+
+      // If there's a new event which matches a request that was previously in
+      // flight, update the associated HttpRequestData.
+      if (outstandingRequestsMap.containsKey(requestId)) {
+        final outstandingRequest = outstandingRequestsMap[requestId];
+        outstandingRequest.merge(requestData);
+        if (!outstandingRequest.inProgress) {
+          outstandingRequestsMap.remove(requestId);
+        }
+        continue;
+      } else if (requestData.inProgress) {
+        outstandingRequestsMap.putIfAbsent(requestId, () => requestData);
+      }
+      currentValues.add(requestData);
+    }
+    // Trigger refresh.
+    _httpRequestsNotifier.value = HttpRequests(
+        requests: currentValues, outstanding: outstandingRequestsMap);
+  }
+
+  void _startPolling() {
+    Future<void>.delayed(const Duration(milliseconds: 1000)).then((_) {
+      if (_httpRecordingNotifier.value) {
+        refreshRequests();
+        _startPolling();
+      }
+    });
+  }
+
+  Future<void> _forEachIsolate(Future Function(IsolateRef) callback) async {
+    final vm = await serviceManager.service.getVM();
+    final futures = <Future>[];
+    for (final isolate in vm.isolates) {
+      futures.add(callback(isolate));
+    }
+    await Future.wait(futures);
+  }
+
+  Future<void> _setHttpTimelineRecording(bool state) async {
+    await _forEachIsolate((isolate) async {
+      final future = serviceManager.service
+          .setHttpEnableTimelineLogging(isolate.id, state);
+      // If the isolate is paused the request above will never complete.
+      await Future.any([
```

Comment:
What should we expect to do when this call fails?

Reply:
I'm still not sure what the best action to take here is. It seems silly to abort recording if we can't enable it on a single isolate, but maybe it would be best to notify the user through a warning dialog or something similar?

Reply:
in practice I expect it should never fail on any isolates unless something quite odd has gone wrong. Given that assumption, I recommend just displaying a  warning toast message and aborting the request to record.
It is better to make it clear to . our users that something went wrong than to mask that fact. If we mask that something went wrong, users will just assume our tools are unreliable.

Reply:
Now that I've thought about this some more, I don't think we actually need to do anything here other than allow for a timeout.

Basically, since `setHttpEnableTimelineLogging` is a `dart:io` extension RPC, if an isolate is paused it won't be able to process the request. However, once the isolate resumes, all outstanding extension requests will be processed. In this case, HTTP request logging being enabled soon after the isolate is resumed as this RPC can never error. AFAIK these requests are handled in a FIFO order, so we won't have to worry about logging being enabled and then later disabled before the isolate is resumed.